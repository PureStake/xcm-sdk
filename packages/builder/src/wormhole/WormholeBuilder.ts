/* eslint-disable @typescript-eslint/no-use-before-define */
import { Wormhole } from '@wormhole-foundation/sdk-connect';
import { EvmParachain, Parachain } from '@moonbeam-network/xcm-types';
import { stringToU8a } from '@polkadot/util';
import {
  WormholeConfigBuilder,
  WormholeConfigBuilderPrams,
} from './WormholeBuilder.interfaces';
import { WormholeConfig } from './WormholeConfig';
import { wormholeFactory } from './wormholeFactory';
import { getExtrinsicAccount } from '../extrinsic/ExtrinsicBuilder.utils';
import { evmToAddress } from '@polkadot/util-crypto/address';

export function WormholeBuilder() {
  return {
    tokenTransfer: ({
      isAutomatic,
    }: {
      isAutomatic: boolean;
    }): WormholeConfigBuilder => ({
      build: ({
        asset,
        destination,
        destinationAddress,
        moonChain,
        source,
        sourceAddress,
      }) => {
        const wh = wormholeFactory(source);
        const whSource = wh.getChain(source.getWormholeName());
        const whDestination = wh.getChain(destination.getWormholeName());

        if (!asset.address) {
          throw new Error('Asset address is required');
        }

        const whAsset = Wormhole.tokenId(whSource.chain, asset.address);
        const whSourceAddress = Wormhole.chainAddress(
          whSource.chain,
          sourceAddress,
        );
        const whDestinationAddress = Wormhole.chainAddress(
          whDestination.chain,
          destinationAddress,
        );

        return new WormholeConfig({
          args: [
            whAsset,
            asset.amount,
            whSourceAddress,
            whDestinationAddress,
            isAutomatic,
            getPayload({ destination, destinationAddress, moonChain }),
          ],
          func: 'tokenTransfer',
        });
      },
    }),
  };
}

export function getPayload({
  destination,
  destinationAddress,
  moonChain,
}: Pick<
  WormholeConfigBuilderPrams,
  'destination' | 'destinationAddress' | 'moonChain'
>): Uint8Array | undefined {
  if (destination.isEqual(moonChain)) {
    return undefined;
  }

  if (!Parachain.is(destination) || !EvmParachain.is(destination)) {
    throw new Error(
      `Destination ${destination.name} is not a Parachain or EvmParachain`,
    );
  }

  // TODO: This is workaround for Peaq EVM. Can we improve it so we don't need to have hardcoded values?
  const isPeaqEvm =
    destination.key === 'peaq-evm-Alphanet' || destination.key === 'peaq-evm';

  return stringToU8a(
    JSON.stringify({
      destination: {
        V3: {
          parents: 1,
          interior: {
            X2: [
              {
                Parachain: destination.parachainId,
              },
              getExtrinsicAccount(
                isPeaqEvm
                  ? evmToAddress(destinationAddress)
                  : destinationAddress,
              ),
            ],
          },
        },
      },
    }),
  );
}
