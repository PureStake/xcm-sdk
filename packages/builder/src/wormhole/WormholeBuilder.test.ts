/* eslint-disable @typescript-eslint/no-explicit-any */
// eslint-disable-next-line import/no-extraneous-dependencies
import { describe, expect, it } from 'vitest';

import { getPolkadotApi } from '@moonbeam-network/xcm-utils';
import {
  wormholeConfigBuilderPrams,
  wormholeToMoonchainConfigBuilderPrams,
} from '../../fixtures';
import { WormholeBuilder } from './WormholeBuilder';

describe('wormholeBuilder', async () => {
  const moonApi = await getPolkadotApi(
    'wss://wss.api.moonbase.moonbeam.network',
  );

  describe('tokenTransfer with isAutomatic=true', () => {
    const transfer = WormholeBuilder().tokenTransfer({ isAutomatic: true });

    it('should be correct config', () => {
      expect(
        transfer.build({ ...wormholeConfigBuilderPrams, moonApi }),
      ).toMatchSnapshot();
    });
  });

  describe('tokenTransfer with isAutomatic=false', () => {
    const transfer = WormholeBuilder().tokenTransfer({ isAutomatic: false });

    it('should be correct config to moon chain', () => {
      expect(
        transfer.build({ ...wormholeToMoonchainConfigBuilderPrams, moonApi }),
      ).toMatchSnapshot();
    });
  });
});
