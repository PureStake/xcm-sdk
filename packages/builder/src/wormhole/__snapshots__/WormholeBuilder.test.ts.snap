// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`wormholeBuilder > tokenTransfer with isAutomatic=false > should be correct config to moon chain 1`] = `
WormholeConfig {
  "args": [
    {
      "address": EvmAddress {
        "address": "0x98891e5FD24Ef33A488A47101F65D212Ff6E650E",
        "type": "Native",
      },
      "chain": "Fantom",
    },
    99000000000n,
    {
      "address": EvmAddress {
        "address": "0xeF46c7649270C912704fB09B75097f6E32208b85",
        "type": "Native",
      },
      "chain": "Fantom",
    },
    {
      "address": EvmAddress {
        "address": "0x98891e5FD24Ef33A488A47101F65D212Ff6E650E",
        "type": "Native",
      },
      "chain": "Moonbeam",
    },
    false,
    undefined,
  ],
  "func": "tokenTransfer",
}
`;

exports[`wormholeBuilder > tokenTransfer with isAutomatic=true > should be correct config 1`] = `
WormholeConfig {
  "args": [
    {
      "address": EvmAddress {
        "address": "0x98891e5FD24Ef33A488A47101F65D212Ff6E650E",
        "type": "Native",
      },
      "chain": "Fantom",
    },
    99000000000n,
    {
      "address": EvmAddress {
        "address": "0xeF46c7649270C912704fB09B75097f6E32208b85",
        "type": "Native",
      },
      "chain": "Fantom",
    },
    {
      "address": EvmAddress {
        "address": "0x0000000000000000000000000000000000000816",
        "type": "Native",
      },
      "chain": "Moonbeam",
    },
    true,
    Uint8Array [
      123,
      34,
      100,
      101,
      115,
      116,
      105,
      110,
      97,
      116,
      105,
      111,
      110,
      34,
      58,
      123,
      34,
      86,
      51,
      34,
      58,
      123,
      34,
      112,
      97,
      114,
      101,
      110,
      116,
      115,
      34,
      58,
      49,
      44,
      34,
      105,
      110,
      116,
      101,
      114,
      105,
      111,
      114,
      34,
      58,
      123,
      34,
      88,
      50,
      34,
      58,
      91,
      123,
      34,
      80,
      97,
      114,
      97,
      99,
      104,
      97,
      105,
      110,
      34,
      58,
      49,
      48,
      48,
      49,
      125,
      44,
      123,
      34,
      65,
      99,
      99,
      111,
      117,
      110,
      116,
      73,
      100,
      51,
      50,
      34,
      58,
      123,
      34,
      105,
      100,
      34,
      58,
      34,
      48,
      120,
      100,
      52,
      51,
      53,
      57,
      51,
      99,
      55,
      49,
      53,
      102,
      100,
      100,
      51,
      49,
      99,
      54,
      49,
      49,
      52,
      49,
      97,
      98,
      100,
      48,
      52,
      97,
      57,
      57,
      102,
      100,
      54,
      56,
      50,
      50,
      99,
      56,
      53,
      53,
      56,
      56,
      53,
      52,
      99,
      99,
      100,
      101,
      51,
      57,
      97,
      53,
      54,
      56,
      52,
      101,
      55,
      97,
      53,
      54,
      100,
      97,
      50,
      55,
      100,
      34,
      44,
      34,
      110,
      101,
      116,
      119,
      111,
      114,
      107,
      34,
      58,
      110,
      117,
      108,
      108,
      125,
      125,
      93,
      125,
      125,
      125,
      125,
    ],
  ],
  "func": "tokenTransfer",
}
`;
